### ISSUE TITLE
Support for BoundServiceAccountTokenVolume refresh token

### ISSUE THREAD

**Issue by hugomcfonseca** [influxdata/telegraf#11267]
## Feature Request

Support to automatically refresh token of the Kubernetes ServiceAccount to avoid hitting authentication issues related to expired ServiceAccount token.

### Proposal:

In theory, using the latest version of [kubernetes/client-go](https://github.com/kubernetes/client-go) should resolve this issue. More info here: https://docs.aws.amazon.com/eks/latest/userguide/service-accounts.html#service-account-tokens.

### Use case:

https://docs.aws.amazon.com/eks/latest/userguide/service-accounts.html#service-account-tokens

### Additional info:

- Cloud Provider: AWS (EKS)
- K8s version: 1.21

For authentication in Kubernetes API, we use the following:

```sh
> k get cm -n monitoring telegraf-daemonset -o yaml | rg -A3 "inputs.kubernetes"
    [[inputs.kubernetes]]
      url = "https://$HOSTIP:10250"
      bearer_token = "/run/secrets/kubernetes.io/serviceaccount/token"
      insecure_skip_verify = true
```

**Comment by powersj** [comment]
Thanks for filing this! 

We are going to need to do some investigation as to what needs to change in Telegraf. We currently pull in `k8s.io/client-go` v0.23.3, so I'm unclear if there is something else we need to do or be doing differently.

**Comment by hugomcfonseca** [comment]
@powersj is there any news on this topic? EKS 1.20 is getting closer to the AWS EOL, and EKS 1.21 will be sooner the minimum version supported by AWS.

**Comment by iliars29** [comment]
I wrote a simple liveness script that checks if token has renewed, if renewed launch telegraf reload with "pkill -1".

That fixed the problem for me.

**Comment by powersj** [comment]
@hugomcfonseca can you share the error message you get from Telegraf so I know where it is erroring out and what type of error you get? I think it is the call to `gathePodInfo` but I want to be sure and see what error comes back.

If it is specific to auth, we could update our logic to re-read the file.

Thanks

**Comment by hugomcfonseca** [comment]
@powersj as of today, I do not have an error on telegraf logs regarding this. Since we use EKS 1.21, it will require that telegraf pods lasts at least 90 days so that it (theoretically) returns an auth error, and that's because of an AWS-specific implementation that extends the lifetime of the ServiceAccount token as it stands:

> For Amazon EKS clusters, the extended expiry period is 90 days. Your Amazon EKS cluster's Kubernetes API server rejects requests with tokens older than 90 days

ref: https://docs.aws.amazon.com/eks/latest/userguide/service-accounts.html#service-account-tokens

I was able to check that token is renewed every hour in the telegraf pods, and check the associated `iss` and `exp` fields. To obtain the token, I just ran:

```sh
kubectl exec -it -n monitoring <telegraf_podname> -- cat /var/run/secrets/kubernetes.io/serviceaccount/token
```

And then, just decode it at https://jwt.io/.

**Comment by powersj** [comment]
For `kube_inventory`, telegraf reads the file during [init](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/kube_inventory/kube_inventory.go#L57) when the plugin is first created as you point out. The token is passed as a string to the [k8s client](https://github.com/influxdata/telegraf/blob/9f3a7414a901deb4d92e81457a7a0570e2342cb9/plugins/inputs/kube_inventory/client.go#L23) via `BearerToken` when the client is also created during init.

Looking at the [k8s config code](https://github.com/kubernetes/client-go/blob/4154d814be868b4bcf610332288a913fca8f4dc1/rest/config.go#L70-L78) it makes it clear, that tokens passed in directly are *not* attempted to be refreshed. While passing in a token file is periodically read.

My proposal then is:

1) If the Telegraf config option `BearerToken` is set then we create the K8s client using `BearerTokenFile` rather than reading the file in Telegraf
2) If the Telegraf config option `BearerTokenString` is set, create the client using `BearerToken` and print an info level log message that no attempt to refresh the token will be made
3) Deprecate the Telegraf config option `BearerTokenString` option 
4) If neither are set, we continue to use the default serviceaccount location, but create the K8s client using `BearerTokenFile`
5) Update the config file with an explanation about refresh requiring a file location

The same steps will be required for both `kubernetes` and `kube_inventory` plugins.

The `prometehus` plugin may also need updates of somesort as there is a similar config option there, but let's see if this even works here first.

**Comment by powersj** [comment]
@hugomcfonseca, or anyone else watching this issue,

I have put up PR #11577 with the proposed changes to the `kube_inventory` plugin. Would it be possible to have you try the artifacts that are attached to that PR in a comment from the Telegraf Tiger?

Thanks!

**Comment by hugomcfonseca** [comment]
@powersj I will test, until tomorrow I will surely have some feedback about it

**Comment by hugomcfonseca** [comment]
@powersj confirmed to be working as expected, I got no more messages in cloudwatch regarding the usage of an expired token! 🚀

**Comment by powersj** [comment]
Thank you for trying the PR so quickly! I'll work on resolving the review questions and land this.

**Comment by hugomcfonseca** [comment]
@powersj by the way, when can we expect a new telegraf version that comes up with this fix?

**Comment by powersj** [comment]
I marked it as a bug fix, so it should come in our final bug fix release of v1.23.4 on ~August 15. If we skip that release, or it does not get included then at the latest, it would come in our next minor release, v1.24.0 ~September 12.
